src/chabochethermal.jl

```diff
@@ -0,0 +1,1050 @@
+- `nu`: Poisson's ratio
+- `alpha`: linear thermal expansion coefficient
+- `R0`: initial yield strength
+- `tvp`: viscoplastic pseudo-relaxation-time (has the units of time)
+- `Kn`: drag stress (has the units of stress)
+- `nn`: Norton-Bailey power law exponent
+- `C1`, `D1`: parameters governing behavior of backstress X1.
+  C1 has the units of stress; D1 is dimensionless.
+- `C2`, `D2`: parameters governing behavior of backstress X2.
+- `C3`, `D3`: parameters governing behavior of backstress X3.
+- `Q`: isotropic hardening saturation state (has the units of stress)
+- `b`: rate of convergence to isotropic hardening saturation (dimensionless)
+"""
+@with_kw struct GenericChabocheThermalParameterState{T <: Real} <: AbstractMaterialState
+    theta0::T = zero(T)  # reference temperature for thermal behavior
+    # basic material parameters
+    E::Function = (theta::Real -> zero(T))
+    nu::Function = (theta::Real -> zero(T))
+    alpha::Function = (theta::Real -> zero(T))
+    R0::Function = (theta::Real -> zero(T))
+    # parameters for viscoplastic overstress model
+    tvp::T = zero(T)
+    Kn::Function = (theta::Real -> zero(T))
+    nn::Function = (theta::Real -> zero(T))
+    # kinematic hardening parameters
+    C1::Function = (theta::Real -> zero(T))
+    D1::Function = (theta::Real -> zero(T))
+    C2::Function = (theta::Real -> zero(T))
+    D2::Function = (theta::Real -> zero(T))
+    C3::Function = (theta::Real -> zero(T))
+    D3::Function = (theta::Real -> zero(T))
+    # isotropic hardening parameters
+    Q::Function = (theta::Real -> zero(T))
+    b::Function = (theta::Real -> zero(T))
+end
+
+"""Problem state for ChabocheThermal material.
+
+- `stress`: stress tensor
+- `R`: yield strength (isotropic hardening)
+- `X1`: backstress 1 (kinematic hardening)
+- `X2`: backstress 2 (kinematic hardening)
+- `X3`: backstress 3 (kinematic hardening)
+- `plastic_strain`: plastic part of strain tensor
+- `cumeq`: cumulative equivalent plastic strain (scalar, ≥ 0)
+- `jacobian`: ∂σij/∂εkl (algorithmic)
+
+The other `dXXXdYYY` properties are the algorithmic jacobians for the
+indicated variables.
+
+The elastic and thermal contributions to the strain tensor are not stored.
+To get them:
+
+    θ₀ = ...
+    θ = ...
+    p = material.parameters
+    v = material.variables
+
+    C(θ) = compliance_tensor(p.E, p.nu, θ)
+    elastic_strain = dcontract(C(θ), v.stress)
+
+    thermal_strain = thermal_strain_tensor(p.alpha, θ₀, θ)
+
+Then it holds that:
+
+    material.drivers.strain = elastic_strain + v.plastic_strain + thermal_strain
+"""
+@with_kw struct GenericChabocheThermalVariableState{T <: Real} <: AbstractMaterialState
+    stress::Symm2{T} = zero(Symm2{T})
+    R::T = zero(T)
+    X1::Symm2{T} = zero(Symm2{T})
+    X2::Symm2{T} = zero(Symm2{T})
+    X3::Symm2{T} = zero(Symm2{T})
+    plastic_strain::Symm2{T} = zero(Symm2{T})
+    cumeq::T = zero(T)
+    jacobian::Symm4{T} = zero(Symm4{T})
+    dRdstrain::Symm2{T} = zero(Symm2{T})
```

Comment:
I haven't yet read further but what is the reason for storing all the partial derivatives? This is OK for small models & material model development but might not be reasonable for materials embedded in FEM as the memory requirements per integration point increase rather much

Reply:
The only reason is that I'm still missing critical pieces of context. :P

Because we were already differentiating one of the variables with respect to one of the drivers, there was an obvious opportunity to generalize, thinking that our users might be interested in the complete algorithmic jacobian - now that we're already computing *∂σ/∂ε* in a consistent manner.

But you're right, it's costly. Let's not compute the extra components if we don't need them.

If we want to produce a consistent *∂σ/∂ε*, we can't avoid computing *∂V/∂ε* for all variables *V*, because the implicit differentiation code will produce all *∂V/∂ε* in one go(*). (At the cost of some CPU time, yes.) See the plastic code path in `integrate_material!`. However, we don't need to actually save *∂R/∂ε* and *∂Xⱼ/∂ε*. Also, we don't need to produce *∂V/∂θ* at all, if it's not used.

Lazy computation, as you suggested offline, could be a solution here (at least for *∂V/∂θ*), in case a user wants the sensitivities to facilitate debugging or some unexpected, exotic use case.

(*) In a nutshell: let *r* be the residual, and *x = [σ, R, X₁, X₂, X₃]* the variables vector (marshaled). By autodiffing *r(x)*, we obtain *∂r/∂x*. By inserting a suitable shim function to change the parameterization to *r(ε)*, and autodiffing that shim, we obtain *∂r/∂ε*. Then we apply the implicit differentiation theorem, which in our case says that *∂x/∂ε = -(∂r/∂x)⁻¹ ∂r/∂ε*. The inversion is of course performed as a left division.

The source of the minus sign in front, that puzzled me and Reijo for a while, is now documented in the source code comments in the function `create_nonlinear_system_of_equations`.

So the algorithm produces the sensitivity w.r.t. *ε* for all variables that happen to be stored in the vector *x*.

In effect, this approach will account for any effect the other variables *R, X₁, X₂, X₃* may have on the result for *∂σ/∂ε*, so the code now produces the actual algorithmic jacobian (up to arithmetic precision).